crates/cli/src/main.rs

```diff
@@ -1,56 +1,238 @@
-use anyhow::{anyhow, bail, Result};
+use anyhow::{anyhow, bail, Context, Result};
 use clap::Parser;
 use core::fmt::Write;
 use std::{
     ffi::OsStr,
     fs,
+    net::SocketAddr,
     path::{Path, PathBuf},
 };
 use wasmi::{
     core::{Value, ValueType, F32, F64},
+    Engine,
     ExportItemKind,
     Func,
     FuncType,
+    Linker,
+    Module,
     Store,
 };
+use wasmi_wasi::{ambient_authority, Dir, TcpListener, WasiCtx, WasiCtxBuilder};
+
+#[derive(Debug, Clone)]
+struct KeyValue {
+    key: String,
+    value: String,
+}
+
+fn parse_env(s: &str) -> Result<KeyValue> {
+    let pos = s
+        .find('=')
+        .ok_or_else(|| anyhow::anyhow!("invalid KEY=value: no `=` found in `{}`", s))?;
+    Ok(KeyValue {
+        key: s[..pos].to_string(),
+        value: s[pos + 1..].to_string(),
+    })
+}
 
 /// Simple program to greet a person
 #[derive(Parser, Debug)]
 #[clap(author, version, about, long_about = None)]
 struct Args {
-    /// The WebAssembly file to execute.
-    #[clap(value_hint = clap::ValueHint::FilePath)]
-    wasm_file: PathBuf,
+    /// The host directory to pre-open for the `guest` to use.
+    #[clap(long = "dir", value_name = "DIR", action = clap::ArgAction::Append, value_hint = clap::ValueHint::DirPath)]
+    dirs: Vec<PathBuf>,
+
+    /// The socket address provided to the module. Allows it to perform socket-related `WASI` ops.
+    #[clap(long = "tcplisten", value_name = "SOCKET_ADDR", action = clap::ArgAction::Append)]
+    tcplisten: Vec<SocketAddr>,
+
+    /// The environment variable pair made available for the program.
+    #[clap(long = "env", value_name = "ENV_VAR", value_parser(parse_env), action = clap::ArgAction::Append )]
+    envs: Vec<KeyValue>,
 
     /// The exported name of the Wasm function to call.
-    ///
-    /// If this argument is missing the wasmi CLI will print out all
+    /// If this argument is missing, the wasmi CLI will try to invoke `""` or `"_start"`
+    /// If those two are not defimed in the module, it prints out all(if any) exported functions
+    /// in the module and ends with an error
     /// exported functions and their parameters of the given Wasm module.
-    #[clap()]
-    func_name: Option<String>,
+    #[clap(long, value_parser)]
+    invoke: Option<String>,
```

Comment:
The `invoke` flag needs to take a custom value of `InvokeArgs` which consist of the function name and its arguments since it does not make sense to have function arguments without an associated function name. The default `_start` function in WASI never takes arguments.
```rust
struct InvokeArgs {
    func_name: String,
    func_args: Vec<String>,
}
```
As with the `KeyValue` type we also need a parse function for this type that is going to be used.
Maybe it is possible to simply derive `clap::Parse` instead of providing a custom `parse_invoke_args` function.

Reply:
Cool. Will Address it.

Reply:
So if `clap::Parse` derive work we should be fine with:
```rust
#[derive(Parser)]
struct InvokeArgs {
    #[clap(value_name = "FUNC_NAME")]
    func_name: Option<String>,
    #[clap()]
    func_args: Vec<String>,
}
```

Reply:
> The default _start function in WASI never takes arguments.

Something to consider: A user might name their exported function `_start` o ` ` and invoke with arguments.

Reply:
`--invoke _start` and `--invoke " "` may be working for those cases, right?

Reply:
Does the code now work so that function arguments can only be provided together with `--invoke` and a function name?

Reply:
Yes. Both `" "` and `_start` can take arguments. 

> Does the code now work so that function arguments can only be provided together with --invoke and a function name?

No. I just understood you. Would have to use a custom parser or use groups. I'll look into it. Maybe I don't need the new struct though. Seen the use of groups in `nextest` cli.

Reply:
From what I know parse groups are usually used when excluding certain options. I think the most straight forward approach is to have `InvokeArgs` as a custom type and simply support it being a CLI flag such as `PathBuf` or `SocketAddr`. I am sure `clap` supports custom types somehow.